### PR TITLE
Add `Axes.get_tick_params()` method.

### PR DESCRIPTION
**Pull request by stefmolin**
## PR Summary
Addresses https://github.com/matplotlib/matplotlib/issues/23603

- Added `Axis.get_tick_params()` method, which will return `Axis._major_tick_kw` or `Axis._minor_tick_kw` depending on the value of `which`
- Added `Axes.get_tick_params()` method, which will return the result of `Axis.get_tick_params()` on the `axis` specified and error out if `both` is passed in

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [X] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by oscargus** [comment]
Based on https://github.com/matplotlib/matplotlib/issues/23633 one may also opt for skipping the `Axes` method and only keep the `Axis` method. This would avoid any mplot3d issues and keep the API a bit cleaner. (But there is also the discoverability drawback of the `Axis` API.)

**Comment by timhoffm** [comment]
@stefmolin sorry for letting this slip.

I'm tempted to only add `Axis.get_tick_params()` for now.

As opposed to the setter, which can set both Axis, and thus can save code, the getter has to resolve to exactly one axis and thus `Axes.get_tick_params()` is just a 1:1 rewriting of the axis access. We can always add the `Axes` method later if we consider it reasonable, but adding it now and taking away later is problematic. So, let's be defensive to start with.

We should cross-ref in the docs of `Axes.tick_params()` and `Axis.tick_params()` which should reduce the discoverability issue a bit.

A very central part of this, is communicating the meaning of these parameters: They are the default values, i.e. if new ticks are created (e.g. by panning or zooming) these values are used for the new ticks. This is in contrast with the actual tick values (the properties of the current tick objects), which can in principle be overwritten by the user.
I think we don't have very clear communication on that right now (in particular also for the setter case `tick_params()` vs. `for tick in ...get_major_ticks(): tick.set_*()`), and it would be good to improve that, but maybe that's going a bit far for this PR and "Get the appearance of ticks, tick labels, and gridlines." is a good enough description for a start.

**Comment by stefmolin** [comment]
> I'm tempted to only add `Axis.get_tick_params()` for now.
> 
> As opposed to the setter, which can set both Axis, and thus can save code, the getter has to resolve to exactly one axis and thus `Axes.get_tick_params()` is just a 1:1 rewriting of the axis access. We can always add the `Axes` method later if we consider it reasonable, but adding it now and taking away later is problematic. So, let's be defensive to start with.

Works for me; I removed the extra method.

> We should cross-ref in the docs of `Axes.tick_params()` and `Axis.tick_params()` which should reduce the discoverability issue a bit.
>
> A very central part of this, is communicating the meaning of these parameters: They are the default values, i.e. if new ticks are created (e.g. by panning or zooming) these values are used for the new ticks. This is in contrast with the actual tick values (the properties of the current tick objects), which can in principle be overwritten by the user.

I updated the docstrings for this &ndash; let me know what you think.

> I think we don't have very clear communication on that right now (in particular also for the setter case `tick_params()` vs. `for tick in ...get_major_ticks(): tick.set_*()`), and it would be good to improve that, but maybe that's going a bit far for this PR and "Get the appearance of ticks, tick labels, and gridlines." is a good enough description for a start.

Not sure what you mean by this. Are you referring to updating a docstring or code?

**Comment by timhoffm** [comment]
> > I think we don't have very clear communication on that right now (in particular also for the setter case `tick_params()` vs. `for tick in ...get_major_ticks(): tick.set_*()`), and it would be good to improve that, but maybe that's going a bit far for this PR and "Get the appearance of ticks, tick labels, and gridlines." is a good enough description for a start.
> 
> Not sure what you mean by this. Are you referring to updating a docstring or code?

The larger point is that we are not clear in distinguishing between the concept of a Axis-wide default tick style (stored in `Axis._major/minor_tick_kw`) and the styles of the actual ticks (stored in the individual tick instances). While it's debatable whether such a distinction is desirable, this is how our current implementation works and it shines through in parts of the API, which sometimes causes confusion to the users.

I'm not sure whether documentation or code is the the right way forward. With documentation, we can explain everything in detail. OTOH we could also try to move the user away from interacting with or thinking about single ticks (#23372), which might make a detailed explanation of the topic unnecessary. This would be part code part documentation.

On a side note, the latter would also be in line with ideas for internal refactoring that should increase performance (https://github.com/matplotlib/matplotlib/issues/5665#issuecomment-875155264).

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -3326,7 +3326,8 @@ def tick_params(self, axis='both', **kwargs):
         Change the appearance of ticks, tick labels, and gridlines.
 
         Tick properties that are not explicitly set using the keyword
-        arguments remain unchanged unless *reset* is True.
+        arguments remain unchanged unless *reset* is True. For the current
+        style defaults, see :meth:`matplotlib.axis.Axis.get_tick_params`.
```
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py:3330]:
```suggestion
        style defaults, see `.Axis.get_tick_params`.
```
we can shorten this, which makes the plain docstring more readable. See also https://matplotlib.org/devdocs/devel/documenting_mpl.html#referring-to-other-code.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -940,6 +940,12 @@ def set_tick_params(self, which='major', reset=False, **kwargs):
 
         For documentation of keyword arguments, see
         :meth:`matplotlib.axes.Axes.tick_params`.
+
+        See Also
+        --------
+        :meth:`.Axis.get_tick_params`
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:946]:
```suggestion
        .Axis.get_tick_params
```
numpydoc does not need markup here (https://numpydoc.readthedocs.io/en/latest/format.html#see-also).

File: lib/matplotlib/axis.py [hunk 3]
```diff
@@ -973,6 +979,28 @@ def set_tick_params(self, which='major', reset=False, **kwargs):
...
 
+    def get_tick_params(self, which='major'):
+        """
+        Get appearance parameters for ticks, ticklabels, and gridlines.
+
+        Parameters
+        ----------
+        which : {'major', 'minor'}, default: 'major'
+            The group of ticks to which the parameters are applied.
+
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:990]: We should add a `Returns` section, stating that it returns a dict, but only contains entries that deviate from the default.

**Comment by timhoffm** [comment]
One additional complication I just noted is that the internal names in `_major/minor_tick_kw` do not match the `tick_params()` keywords:

https://github.com/matplotlib/matplotlib/blob/3b79f63980c767fabc59ec5ba267818d8b4b5671/lib/matplotlib/axis.py#L948

We have to either translate back for `get_tick_params()` or clearly document that you get something else :frowning:

**Comment by stefmolin** [comment]
> We have to either translate back for `get_tick_params()` or clearly document that you get something else 😦

@timhoffm &ndash; Is there a preference? I can see these being confusing if you get them back:

https://github.com/matplotlib/matplotlib/blob/3b79f63980c767fabc59ec5ba267818d8b4b5671/lib/matplotlib/axis.py#L1009-L1016

The rest are probably fine. What do you think?

**Comment by timhoffm** [comment]
Uh, this is a difficult decision. I'm tempted to translate back. I think the docstring https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.tick_params.html#matplotlib.axes.Axes.tick_params is the only user-visible place where we document what can go in (we currently accept more, but that somewhat coincidence). In that sense, `get_tick_params` should return the names as given there.

This back-and forth converting is a bit annoying internally, but I think it is the most logical and consistent interface we can offer here.

**Comment by stefmolin** [comment]
I added a reverse option to the `_translate_tick_params()` method and updated it to no longer alter the input dictionary.  I'm using this in the forward and reverse directions in the test, so we can make sure both directions work going forward.

File: lib/matplotlib/axes/_base.py (hunk 1 above)
Comment (stefmolin) [reply lib/matplotlib/axes/_base.py]: This doesn't seem to work in the [generated docs](https://output.circle-artifacts.com/output/job/58a01a40-e3df-4d10-a3a1-57366c92e770/artifacts/0/doc/build/html/api/_as_gen/matplotlib.axes.Axes.tick_params.html#matplotlib.axes.Axes.tick_params).

**Review by timhoffm — APPROVED** [review]

**Comment by stefmolin** [comment]
@timhoffm - Any ideas what is going on for the docs per [this comment](https://github.com/matplotlib/matplotlib/pull/23692/#discussion_r1002745295) above?

Some additional questions on documentation for this PR:

1. Do we need an example to the docs for this?
2. Should I add `get_tick_params()` to the What's New section?
3. Since `_translate_tick_params()` no longer alters the keyword dict, should we add it to the API changes? Or does this not matter since it is not for external use?

**Comment by timhoffm** [comment]
> @timhoffm - Any ideas what is going on for the docs per [this comment](https://github.com/matplotlib/matplotlib/pull/23692/#discussion_r1002745295) above?

I'l check later.

>     1. Do we need an example to the docs for this?

You can try and add an Example section to the docstring. I would not make an addition to the Example Gallery.

I hope this is correct, because it'S only from memory - please ignore if I'm wrong: It may be a bit confusing because AFAIR there are a few entries in _major_tick_kw by default. But we're stating that this only contains deviations from defaults (having "deviations from defaults" by default is a bit awkward). OTOH it may make sense to spell this out.

>     2. Should I add `get_tick_params()` to the What's New section?

Yes, please.
 
>     3. Since `_translate_tick_params()` no longer alters the keyword dict, should we add it to the API changes? Or does this not matter since it is not for external use?

No API change note, because our public API has not changed.

**Comment by timhoffm** [comment]
Just close the PR. Open PRs anticipate a follow up action on the PR, even if that action is only "decide what to do".

**Comment by stefmolin** [comment]
> Just close the PR. Open PRs anticipate a follow up action on the PR, even if that action is only "decide what to do".

@timhoffm What do you mean by this? Was this meant for another PR?

**Comment by jklymak** [comment]
I'm 98% sure @timhoffm meant this for #23787

**Comment by timhoffm** [comment]
Yes, sorry! :sheep:

**Comment by stefmolin** [comment]
> I hope this is correct, because it'S only from memory - please ignore if I'm wrong: It may be a bit confusing because AFAIR there are a few entries in _major_tick_kw by default. But we're stating that this only contains deviations from defaults (having "deviations from defaults" by default is a bit awkward). OTOH it may make sense to spell this out.

Yeah, there are a few entries in there by default. I agree the phrasing is a little confusing. I modified the docstrings to indicate that we are returning the appearance of any new elements that will be added without calling out defaults – hopefully, that makes it easier to understand.

I also added the example to the docstring and what's new entry.

**Review by timhoffm — APPROVED** [review]

File: doc/users/next_whats_new/view_current_axis_format.rst [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+.. code-block:: python
+
+    import matplotlib.pyplot as plt
+
+    fig, ax = plt.subplots()
+    ax.yaxis.set_tick_params(labelsize=30, labelcolor='red',
+                             direction='out', which='major')
+    print(ax.yaxis.get_tick_params(which='major'))
+    print(ax.yaxis.get_tick_params(which='minor'))
```
Comment (jklymak) [inline doc/users/next_whats_new/view_current_axis_format.rst:16]: This code block doesn't really tell me too much by itself.  Can it get the result of the print statements added as well?

File: lib/matplotlib/axes/_base.py [hunk 5]
```diff
@@ -3403,7 +3403,8 @@ def tick_params(self, axis='both', **kwargs):
         Change the appearance of ticks, tick labels, and gridlines.
 
         Tick properties that are not explicitly set using the keyword
-        arguments remain unchanged unless *reset* is True.
+        arguments remain unchanged unless *reset* is True. For the current
+        style defaults, see `.Axis.get_tick_params`.
```
Comment (jklymak) [inline lib/matplotlib/axes/_base.py:3407]: I find this new clause somewhat ambiguous.  Do you really mean "defaults"?  I think maybe this should be "For the current tick parameters see `.Axis.get_tick_params`." if this is needed at all.

File: lib/matplotlib/axis.py [hunk 6]
```diff
@@ -916,6 +916,12 @@ def set_tick_params(self, which='major', reset=False, **kwargs):
 
         For documentation of keyword arguments, see
         :meth:`matplotlib.axes.Axes.tick_params`.
+
+        See Also
+        --------
+        .Axis.get_tick_params
+            View the current style defaults for ticks, ticklabels, and
```
Comment (jklymak) [inline lib/matplotlib/axis.py:923]: Again, I don't think these are necessarily the defaults, are they?  They are the current settings...

File: lib/matplotlib/axis.py [hunk 7]
```diff
@@ -949,8 +955,62 @@ def set_tick_params(self, which='major', reset=False, **kwargs):
...
+
+        Parameters
+        ----------
+        which : {'major', 'minor'}, default: 'major'
+            The group of ticks for which the parameters are retrieved.
+
+        Returns
+        -------
+        dict
+            Properties for styling *new* elements added to this axis.
```
Comment (jklymak) [inline lib/matplotlib/axis.py:970]: OK, reading this I finally understand what you mean by "default style".  But I think that is incorrect terminology, and more confusing than helpful.  I think if someone has manually set a tick to a different styling, they will know they did it.  
```suggestion
            Properties for styling tick elements added to the axis.
```

File: lib/matplotlib/axis.py [hunk 8]
```diff
@@ -949,8 +955,62 @@ def set_tick_params(self, which='major', reset=False, **kwargs):
...
+
+        Notes
+        -----
+        This method returns the appearance parameters for styling *new*
+        elements added to this axis and may be different from the values
+        on current elements, if they were modified directly by the user
+        (e.g., via ``set_*`` methods on individual tick objects).
```
Comment (jklymak) [inline lib/matplotlib/axis.py:977]:
```suggestion
        on current elements if they were modified directly by the user
        (e.g., via ``set_*`` methods on individual tick objects).
```

File: lib/matplotlib/axis.py [hunk 9]
```diff
@@ -949,8 +955,62 @@ def set_tick_params(self, which='major', reset=False, **kwargs):
...
+            'labelcolor': 'red'}
+            >>> ax.yaxis.get_tick_params(which='minor')
+            {'left': True,
+            'right': False,
+            'labelleft': True,
+            'labelright': False,
+            'gridOn': False}
+
+        This allows us to confirm without visual inspection of the plot
+        that our styling is having the desired effect.
```
Comment (jklymak) [inline lib/matplotlib/axis.py:1002]:
```suggestion
```

**Review by jklymak — COMMENTED** [review]
Just a few documentation suggestions.  I didn't read the history, but the docs refer to a mysterious "default style" which to me are the rcParam values, not the values currently on the axis.  I think you are trying to differentiate between manual ticks and those styles automatically, but I think the note gets that across adequately.  People don't often much with individual ticks, so I'd argue not to muddy the waters of the main docstrings with a rare case.

**Comment by stefmolin** [comment]
Should I also add the `.. versionadded::` directive per the latest version of the PR template? Would this be for 3.7?

**Comment by stefmolin** [comment]
@jklymak - I added the `.. versionadded::` directive and tweaked the docstrings and example per your comments.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
Thank you @stefmolin !

File: lib/matplotlib/axis.py [hunk 10]
```diff
@@ -988,19 +1051,27 @@ def _translate_tick_params(kw):
             'labelright': 'label2On',
             'labeltop': 'label2On',
```
Comment (GenieTim) [inline lib/matplotlib/axis.py:1052]: labelleft & labelbottom have the same axis key, the translation will therefore always return `labelleft` and `labelright`, also for the x-axis. This is confusing.

Comment (QuLogic) [reply lib/matplotlib/axis.py:1052]: Please open a new issue with the details of what you are seeing.

Comment (GenieTim) [reply lib/matplotlib/axis.py:1052]: Ok, done, see #28772

**Pull request merged**